src/test/java/api/loans/ScheduledDueDateNoticesProcessingTests.java

```diff
@@ -0,0 +1,385 @@
+    InterruptedException,
+    TimeoutException,
+    ExecutionException {
+
+    DateTime justBeforeDueDateTime = dueDate.minusSeconds(1);
+    scheduledNoticeProcessingClient.runNoticesProcessing(justBeforeDueDateTime);
+
+    checkSentNotices(beforeTemplateId);
+
+    checkScheduledNotices(
+      null,
+      dueDate,
+      dueDate.plus(afterPeriod.timePeriod()));
+  }
+
+  @Test
+  public void beforeAndUponAtNoticesShouldBeSentWhenProcessingJustAfterDueDate()
+    throws MalformedURLException,
+    InterruptedException,
+    TimeoutException,
+    ExecutionException {
+
+    DateTime justAfterDueDateTime = dueDate.plusSeconds(1);
+    scheduledNoticeProcessingClient.runNoticesProcessing(justAfterDueDateTime);
+
+    checkSentNotices(beforeTemplateId, uponAtTemplateId);
+
+    checkScheduledNotices(
+      null,
+      null,
+      dueDate.plus(afterPeriod.timePeriod()));
+  }
+
+  @Test
+  public void afterRecurringNoticeShouldBeSentSeveralTimesBeforeLoanIsClosed()
+    throws MalformedURLException,
+    InterruptedException,
+    TimeoutException,
+    ExecutionException {
+
+    DateTime justAfterDueDateTime = dueDate.plusSeconds(1);
+    scheduledNoticeProcessingClient.runNoticesProcessing(justAfterDueDateTime);
+    //Clear all sent notices before actual test
+    patronNoticesClient.deleteAll();
+
+    DateTime afterNoticeRunTime = dueDate.plus(afterPeriod.timePeriod()).plusSeconds(1);
+    scheduledNoticeProcessingClient.runNoticesProcessing(afterNoticeRunTime);
+
+    DateTime afterNoticeExpectedRunTime = dueDate
+      .plus(afterPeriod.timePeriod())
+      .plus(afterRecurringPeriod.timePeriod());
+
+    checkScheduledNotices(
+      null,
+      null,
+      afterNoticeExpectedRunTime);
+
+    //Run again to send recurring notice
+    scheduledNoticeProcessingClient.runNoticesProcessing(
+      afterNoticeExpectedRunTime.plusSeconds(1));
+
+    checkSentNotices(afterTemplateId, afterTemplateId);
+
+    DateTime secondRecurringRunTime =
+      afterNoticeExpectedRunTime.plus(afterRecurringPeriod.timePeriod());
+
+    checkScheduledNotices(
+      null,
+      null,
+      secondRecurringRunTime);
+
+    loansFixture.checkInByBarcode(item);
+    //Clear sent notices again
+    patronNoticesClient.deleteAll();
+
+    //Run after loan is closed
+    scheduledNoticeProcessingClient.runNoticesProcessing(
+      secondRecurringRunTime.plusSeconds(1));
+
+    checkSentNotices();
+    checkScheduledNotices(null, null, null);
+  }
+
+  @Test
+  public void processingTakesNoticesInThePastLimitedAndOrdered()
+    throws MalformedURLException,
+    InterruptedException,
+    TimeoutException,
+    ExecutionException {
+    //Clean scheduled notices before this test
+    scheduledNoticesClient.deleteAll();
+
+    DateTime systemTime = DateTime.now(DateTimeZone.UTC);
+    int expectedNumberOfUnprocessedNoticesInThePast = 10;
+    int numberOfNoticesInTheFuture = 20;
+
+    List<JsonObject> noticesInThePast = IntStream.iterate(0, i -> i + 1)
```

Comment:
This and the below can be extracted to a method something like the below to help reveal the intent. Feel free to rename the method and parameters if you can think of improvements.
```
  private List<JsonObject> createNoticesOverTime(
    Function<Integer, DateTime> timeOffset, int numberOfNotices) {

    return IntStream.iterate(0, i -> i + 1)
      .boxed()
      .map(timeOffset)
      .map(this::createFakeScheduledNotice)
      .limit(numberOfNotices)
      .collect(Collectors.toList());
  }
```

Reply:
Thank you for the suggestion, I've applied this.